### PR TITLE
Improve xHRM 

### PR DESCRIPTION
**Pull request by samveen**
- Fix for #3923 
- General Improvements to script via ShellCheck.

### CODE REVIEW

**Comment by samveen** [comment]
@bybai @immarvin  Please review. I am hoping this makes the `2.13.8` release

**Comment by bybai** [comment]
@samveen , Thanks for your great effort to xHRM, 2.13.8 will be released soon. And we need more time to test xHRM code changes, then merge it. So it is not a good time to make it into 2.13.8 , this commit will be in 2.13.9 release.

**Comment by whowutwut** [comment]
@bybai should we consider merging now?

File: xCAT-server/share/xcat/scripts/xHRM [hunk 1]
```diff
@@ -235,33 +234,19 @@ elif [ "bridgeprereq" = "$1" ]; then
     brctl addbr $BNAME
     brctl setfd $BNAME 0 #fast forwarding
     ip link set $BNAME up
-    saveroutes=$(ip route | grep default| grep "dev $PORTS"|grep via|sed -e 's/dev .*//')
-    OIFS=$IFS
-    IFS=$'\n'
-    saveip=$(ip addr show dev $PORTS scope global|grep inet|sed -e 's/inet.//'|sed -e 's/[^ ]*$//')
+    saveroutes="$(ip route | grep default| grep "dev $PORTS"|grep via|sed -e 's/dev .*//')"
+    saveip="$(ip -4 -o addr show dev enp0s3 scope global | sed 's/.*inet //'| sed 's/\( global \).*/\1/')"
```
Comment (bybai) [inline xCAT-server/share/xcat/scripts/xHRM]: ip -4 -o addr show dev enp0s3 scope global
I think 'enp0s3' in line 238 should be replaced by a variable;

File: xCAT-server/share/xcat/scripts/xHRM [hunk 2]
```diff
@@ -1,6 +1,6 @@
-#!/bin/bash  
-if [ "$(uname -s|tr 'A-Z' 'a-z')" = "linux" ];then
-   str_dir_name=`dirname $0`
+#!/bin/bash -u
```
Comment (bybai) [inline xCAT-server/share/xcat/scripts/xHRM]: Is "-u" needed here?  
We usually do not use it in ``#!/bin/bash`` shebang line.
-u :  Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion. An error message will be written to the standard error, and a non-interactive shell will exit.

**Comment by bybai** [comment]
Hi @whowutwut, the xHRM code logic and functions are not changed, most changes are about "using $() to replace ``" and improvements to script via ShellCheck, we can accept these changes. But "'enp0s3' in line 238 should be replaced by a variable", this should be updated.

**Comment by samveen** [comment]
@bybai Rebased on master, and fixed both the issues: removed `-u` and corrected line `238` to correct hardcoded name with `$PORTS`

File: xCAT-server/share/xcat/scripts/xHRM (hunk 2 above)
Comment (samveen) [reply xCAT-server/share/xcat/scripts/xHRM]: done

File: xCAT-server/share/xcat/scripts/xHRM (hunk 1 above)
Comment (samveen) [reply xCAT-server/share/xcat/scripts/xHRM]: done

**Comment by bybai** [comment]
Thanks @samveen , since there are lots of code changes, let me have a quick dry run, then merge it.

File: xCAT-server/share/xcat/scripts/xHRM [hunk 3]
```diff
@@ -235,33 +234,19 @@ elif [ "bridgeprereq" = "$1" ]; then
...
-    saveip=`ip addr show dev $PORTS scope global|grep inet|sed -e 's/inet.//'|sed -e 's/[^ ]*$//'`
+    saveroutes="$(ip route | grep default| grep "dev $PORTS"|grep via|sed -e 's/dev .*//')"
+    saveip="$(ip -4 -o addr show dev $PORTS scope global | sed 's/.*inet //'| sed 's/\( global \).*/\1/')"
     if [ ! -z "$saveip" ]; then
-        for line in $saveip; do 
-            IFS=$OIFS
-            newline=`echo $line|sed 's/dynamic//g'`
-            ip addr add dev $BNAME $newline
-        done
+        ip addr add dev $BNAME $newline
```
Comment (bybai) [inline xCAT-server/share/xcat/scripts/xHRM]: There are 2 problems:
1, $newline is unbound variable
2, I saw you deleted for loop, if there are several ips in saveip, "ip addr add dev ...." will fail.

File: xCAT-server/share/xcat/scripts/xHRM [hunk 4]
```diff
@@ -235,33 +234,19 @@ elif [ "bridgeprereq" = "$1" ]; then
...
     if [ ! -z "$saveip" ]; then
-        OIFS=$IFS
-        IFS=$'\n'
-        for line in $saveip; do 
-            IFS=$OIFS
-           newline=`echo $line|sed 's/dynamic//g'`
-            ip addr del dev $PORTS $newline
-        done
-       IFS=$OIFS
+        ip addr del dev $PORTS $newline
```
Comment (bybai) [inline xCAT-server/share/xcat/scripts/xHRM]: The same with line 240

**Comment by samveen** [comment]
@bybai Yeah I found the same issue: it's due to the newly added code for #4060 which adds checks for default interface.

I'm refactoring that check to happen IFF correct parameters are not passed.

File: xCAT-server/share/xcat/scripts/xHRM (hunk 4 above)
Comment (samveen) [reply xCAT-server/share/xcat/scripts/xHRM]: Fixed

File: xCAT-server/share/xcat/scripts/xHRM (hunk 3 above)
Comment (samveen) [reply xCAT-server/share/xcat/scripts/xHRM]: fixed

**Comment by bybai** [comment]
Hi @neo954 , could you help try a quick UT for this script?

**Comment by neo954** [comment]
These patches look good. I will merge this one.

**Pull request merged**